### ISSUE TITLE
Input is predicted by server and clients, but the correct/received input isn't used

### ISSUE THREAD

**Issue by TheYellowArchitect** [foxssake/netfox#280]
It is safe to assume the server does a rollback, because it receives inputs from clients and rolls back to the tick of the received input up to the current state.
The below is the prediction/extrapolation the server does
![netfox-input-prediction](https://github.com/user-attachments/assets/e0c359f5-9ba7-41b5-919a-4bc8a973f895)
and it seems input prediction/extrapolation is done not only by server but clients too.
![input-prediction-for-clients-too](https://github.com/user-attachments/assets/40ac671f-75f7-483a-9661-ddb1e482b7ad)

This conflicts an old claim that there is no input prediction and the server just waits for inputs:
![rollback-server-doesnt-extrapolate-or-rollback](https://github.com/user-attachments/assets/8f4114d6-36e4-4d7f-91a1-70585965296b)

The weird thing is that once the correct input arrives, there seems to be no rollback from that input up to latest tick, so the correct input doesn't matter. The new input once it arrives, should trigger a `rollback_tick` and apply itself, but it doesn't. `rollback_tick` runs only once per frame. To prove this, put this code in `brawler-controller.gd`:
```gdscript
var debug_tick_directions: Dictionary = {} #<int, Vector3>

...

func _rollback_tick(delta, tick, is_fresh)
...
    # Place the below under the direction section
    if (multiplayer.is_server() && player_id != 1):
	if (debug_tick_directions.has(tick) == false):
		debug_tick_directions[tick] = direction
	else:
		print("Server: NetworkTime.tick (%s) Rollback Correction for Avatar %s at tick %s directions are: %v | %v" % [NetworkTime.tick, player_id, tick, direction, debug_tick_directions[tick]])
```
I cannot get the above to ever print with 3 players and 200ms delay.

--------

The rollback only happens for the latest `NetworkTime.tick` because of this line, which applies to server and clients regardless of authority: 

```gdscript
func _after_loop():
	_earliest_input = NetworkTime.tick
```

and on `submit_input` rpc function, it goes down for non-authorities (those who receive input RPC):
```gdscript
earliest_input = min(_earliest_input, tick_received)
```

Why focus on the variable `_earliest_input`? Because it is used to determine which tick a rollback starts
![server-rollback-input](https://github.com/user-attachments/assets/85bcda17-eff9-47d4-92c3-376ab7ea7f56)

```gdscript
## Submit the resimulation start tick for the current loop.
##
## This is used to determine the resimulation range during each loop.
func notify_resimulation_start(tick: int):
	_resimulate_from_tick = min(_resimulate_from_tick, tick)
```
![rollback](https://github.com/user-attachments/assets/d9e4b529-afbb-4a55-ae2a-7a86b5fa47ec)

![rollback-spam](https://github.com/user-attachments/assets/ea474de0-5a3e-4df7-82a7-949b9ad5ac03)

From my understanding, assuming all clients take exactly 3 ticks to send an input to the server:
At tick 100 (latest), the server has the clients' inputs of tick 97 and the input ticks of 98, 99, 100, are predictions/extrapolations using the same input of tick 97.
Let's say a client changes input at tick 99, and this input arrives to the server at tick 102. The server does replace its `_inputs[]` but it doesn't rollback to tick 99 up to 102. It rolls back to 101 (see `after_loop` function above) if not just proceeds with 102.

This would mean that the server, whenever it sends the state of any tick, e.g. tick 100, it isn't the true/final state, as it contains predicted/extrapolated inputs. And by extension, the server will rollback many ticks per frame because there is no input delay, which is not the case currently. Or have I missed a good part of the puzzle?

-----

Related to all of the above "rollback doesn't work as expected" is that the state is RPC submitted at `record_tick` which happens within a rollback. So multiple states can be RPC submitted in a single tick if multiple rollbacks happen. This ofc doesn't currently happen, because `rollback_tick` doesn't run more than once per tick (as shown in the code example with `debug_ticks_direction: Dictionary`). So the state submission must be moved to `after_loop` in order to send only the latest state. Unless a new rollback is to also send corrected states? (I would suggest against this, as it kills the network because of bandwidth)

----

Being unable to debug further without spending days, I just want to understand how input prediction works in netfox. Forest brawl works, no one can deny that. But seeing how the bombs are directly injected into the state and are unaffected by rollback, and this general weirdness I discovered above, I feel something is fundamentally wrong, despite it working smoothly even at 200ms in a demo. I want to understand netfox fundamentally so as to expand it for physics or items.

----

P.S. I tried making a [commit](https://github.com/foxssake/netfox/commit/626f6e33eb624704fea1d8169b6f1554a886c2c8#r147025921) which closes this issue, by:
1. Updating the value of `_earliest_input` for input authority, to equal the latest `NetworkTime.tick`
2. Updating the value of `_earliest_input` for **non**-input authorities, to equal the latest `_inputs[]` index received.
3. Moved the state submission so it doesn't trigger per rollback tick, but after each rollback ends. 
4. Debug print for `_rollback_tick` to confirm it triggers multiple times per rollback (it does)

But it bugs a lot, and idk why :cry:

**Comment by elementbound** [comment]
### tl;dr: There's (still) no input prediction in netfox, please read the code again

---

Ok so I want to help you on this, but the way this issue is written makes it really difficult for the following reasons:

1. Please specify the goal of the ticket explicitly
    * For a while, it's about input prediction, then how input is not handled well, then you end with wanting to understand rollback, but then change it to understanding netfox itself
    * Happy to help with either, but please pick one per ticket
1. Please include a direct problem statement
    * This is a long description of a hypothetical about how netfox might work
    * No actual bug is described ( current observed behavior vs. expected behavior )
    * Some quirkiness is discussed, but I'd prefer if you *explicitly* specified - preferably in a dedicated section - what's the expected vs. observed behavior, instead of me having to guesstimate which part you need help with
1. Please include code snippets as markdown code blocks
    * Screenshots are not as easy to read and navigate or even refer to as text
    * Some screenshots are also missing enough context so I have to figure out from memory which function are you referring to
1. Please use the original source code for examples
    * Discerning which part is original netfox code and which is temp / debug / custom gets even more disorienting with the above included
    * This got especially confusing on the first two screenshots, where I see that both refer to RollbackSynchronizer, but e.g. line 227 contains different text on both screenshots

Also please make sure to keep the above in mind in the future as well, when opening issues. Thanks!

---

> This conflicts an old claim that there is no input prediction and the server just waits for inputs

Once again, there's no input prediction nor extrapolation. For your related example, please read the `_can_simulate` method in `RollbackSynchronizer` and refer to how that is integrated into the overall rollback process. I don't exactly follow what's the motivation behind the logs on the first two screenshots, but they don't have much bearing on input prediction - the logs are not on the same level where the input checks are made. 

---

> I cannot get the above to ever print with 3 players and 200ms delay.

That is the expected behavior, since there's no input prediction.

Let's say you're on tick 100 and an input arrives for tick 97. Tick 97 is simulated, then 98, 99, and 100 are not, since we don't have an input for it. On tick 101, input for tick 98 arrives. Tick 98 is simulated, then 99, 100, and 101 is not since we don't have input for it. So no ticks are simulated twice in rollback. 

---

> The rollback only happens for the latest NetworkTime.tick because of this line

False. Please read carefully how the rollback loop is setup, in its entirety, in NetworkRollback's code and its corresponding [documentation](https://foxssake.github.io/netfox/netfox/guides/network-rollback/).

> Let's say a client changes input at tick 99, and this input arrives to the server at tick 102. The server does replace its _inputs[] but it doesn't rollback to tick 99 up to 102. It rolls back to 101 (see after_loop function above) if not just proceeds with 102.

False. Again, please read the documentation and related code on how rollback is handled. 

> And by extension, the server will rollback many ticks per frame because there is no input delay, which is not the case currently.

Given your example, if inputs take 3 ticks to arrive, the server will rollback 3 ticks per *tick loop* ( not frame ), and simulate only the first for the player, as that's the only one the server has input for. 

---

> So the state submission must be moved to after_loop in order to send only the latest state.

False. Please see above why only a single state is submitted as a new input arrives.

---

> I just want to understand how input prediction works in netfox

Again, there is no input prediction in netfox. 

> I feel something is fundamentally wrong

I suggest you familiarize yourself with the codebase and the documentation.

---

Look, I've spent more than an hour reading this through, making sure I don't miss anything, even asked for help to get an extra perspective, and, from my point of view, this is a train of thought based on multiple misunderstandings about network rollback stemming from ( assuming ) not reading critical parts of the code. There is no actual issue I can fix or act upon. 

I'm happy to help with questions, issues, bugs, or even helping with debugging user projects, but so far this issue is just being unfamiliar with how netfox works, despite a whole documentation site, comments in code, and a Discord server for quick questions, **and multiple discussions on how rollback works.**

Anyway, let me know if you have any specific questions!

**Comment by TheYellowArchitect** [comment]
Thank you for the answer. The input prediction clarification specifically helped (I opened https://github.com/foxssake/netfox/issues/299 for a question I have on that) 

To completely clear my confusion and the confusion I caused on you, and because I have to pack my stuff for a long drive in a few hours, I will ask only critical questions:

>That is the expected behavior, since there's no input prediction.
>Let's say you're on tick 100 and an input arrives for tick 97. Tick 97 is simulated, then 98, 99, and 100 are not, since we don't have an input for it. On tick 101, input for tick 98 arrives. Tick 98 is simulated, then 99, 100, and 101 is not since we don't have input for it. So no ticks are simulated twice in rollback.

The above is for the non-input authority. I also confirmed this via debugging.

When we are at tick 100 with the input authority, our avatar is at tick 100 input and state, instead of tick 97's? If yes, this means our avatar is at 100 while the other avatars (which we are not input authority) are at tick 97, correct?

The above example, whatever the answer, must be included in the documentation

>>And by extension, the server will rollback many ticks per frame because there is no input delay, which is not the case currently.
>Given your example, if inputs take 3 ticks to arrive, the server will rollback 3 ticks per tick loop ( not frame ), and simulate only the first for the player, as that's the only one the server has input for.

Could you elaborate on this?  Specifically this

>and simulate only the first for the player

Doing a rushed debugging on this use-case, by adding the following code at the end of `RollbackSynchronizer.submit_input()`
```gdscript
if (multiplayer.is_server() && root.player_id != 1):
		_logger.info("Received input at tick: %s" % tick)
``` and the following code inside `BrawlerController._rollback_tick()`
```gdscript
if (multiplayer.is_server() && player_id != 1):
		_logger.info("This tick is: %s" % tick)
```
and with 200 ms delay, there is only 1 rollback tick per loop - it works as expected from what you wrote.
E.g. at tick 214, the server receives input of tick 205, and at tick 215, runs `_rollback_tick(205)` and receives input of tick 206. At tick 216, the server runs `_rollback_tick(206)`

I will debug this when home to confirm if currently the server can run 3 rollback ticks per loop. because in theory it should work, but haven't tested it. I assume it runs 3 rollback ticks per loop if 3 inputs are missing in a row

**Comment by elementbound** [comment]
> When we are at tick 100 with the input authority, our avatar is at tick 100 input and state, instead of tick 97's? If yes, this means our avatar is at 100 while the other avatars (which we are not input authority) are at tick 97, correct?

Correct. For the local avatar, we have the input for all ticks, the state from the server for ticks up to 97, and we can predict 98, 99, and 100, so state 100 is visible. For the other players' avatars, we can only rely on state, which as only arrived up to tick 97. 

> The above example, whatever the answer, must be included in the documentation

I reserve the right to judge what *must* or *must not* be done for this repo :slightly_smiling_face: I agree nonetheless. My point is, I invite you to ask / propose / suggest things to do, instead of declaring them as must-do's. 

---

> Could you elaborate on this? Specifically this
>> and simulate only the first for the player
> first rollback tick loop? The earlier or the latest?

Staying with the example, we're about to run tick 100 in the network tick loop on the server, and we've received input for tick 97. During the network tick loop for tick 100, the rollback tick loop will simulate ticks 97, 98, 99, and 100. The player will be simulated only for tick 97, since that's the only tick the server has input for. Input for ticks 98, 99 and 100 have not arrived yet. 

---

> I assume it runs 3 rollback ticks per loop if 3 inputs are missing in a row, which would also submit 3 state RPCs.

If the server suddenly receives 3 tick's worth of inputs at once, then yes, this can happen.

**Comment by TheYellowArchitect** [comment]
Thank you for the clarity, I understand better now. Closing this issue because as you said its not really an issue but a misunderstanding from my part because I thought this was a bug.
